### PR TITLE
(CDAP-16369) Remote Execution Runtime Redesign

### PR DESCRIPTION
**Pull request by chtyim**
Changes are organized by commits. The biggest change is in the commit "Change job launching and monitoring mechanism".

In summary this change the the runtime launching and monitoring for non-native profile program execution.

Job launch can be done in two ways.
1. SSH
1. RuntimeJobManager

The job launching mechanism is controlled by provisioner implementation.

Job monitoring and metadata collection are always posted from the remote runtime to the CDAP runtime server. There are two ways that the remote runtime can communicate with CDAP.
1. Through SSH tunnel from CDAP and socks5 proxy that runs in the remote runtime process
1. A direct URL exposed by CDAP with authentication support

The job monitoring mechanism is controlled by cdap-site configuration and can be overridden through runtime arguments 

The job launching and job monitoring mechanism can be mix-and-match.

### CODE REVIEW

File: cdap-runtime-spi/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/RuntimeJob.java [hunk 1]
```diff
@@ -28,4 +28,9 @@
    * @throws Exception thrown if any error while job execution
    */
   void run(RuntimeJobEnvironment runtimeJobEnvironment) throws Exception;
```
Comment (CuriousVini) [inline cdap-runtime-spi/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/RuntimeJob.java:30]: should we call this `start`?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/GceRemoteAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A {@link RemoteAuthenticator} that authenticate remote calls using Google Cloud token acquired from GCE metadata.
+ */
+@SuppressWarnings("unused")
+public class GceRemoteAuthenticator extends RemoteAuthenticator {
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/GceRemoteAuthenticator.java:36]: could you please file a jira to move it to a separate extension and add it as todo here?

**Review by CuriousVini — APPROVED** [review]
2 minor comments. LGTM 👍

**Review by CuriousVini — APPROVED** [review]

File: cdap-runtime-spi/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/RuntimeJob.java (hunk 1 above)
Comment (chtyim) [reply cdap-runtime-spi/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/RuntimeJob.java:30]: changed to requestStop()

**Review by CuriousVini — APPROVED** [review]
lgtm 👍

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/GceRemoteAuthenticator.java (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/GceRemoteAuthenticator.java:36]: Done. CDAP-16583

**Comment by chtyim** [comment]
Build passed https://builds.cask.co/browse/CDAP-DUT7150-6

**Pull request merged**